src/_pytest/fixtures.py

```diff
@@ -232,11 +233,18 @@ def getfixturemarker(obj: object) -> Optional["FixtureFunctionMarker"]:
     )
 
 
-# Parametrized fixture key, helper alias for code below.
-_Key = Tuple[object, ...]
+@dataclasses.dataclass(frozen=True)
+class FixtureArgKey:
+    argname: str
+    param_index: Optional[int]
```

Comment:
param index and value have a behavioral code smell as we have exactly 2 variants of this class now

1. with the index and only the index
2. with the value and only the value

at first glance this looks like a possible change in behavior due to value vs index
but its also not clear to me how to test/validate this in more detail

i do wonder if a namedtuple would make sense and i do wonder if using 2 types would make sense to avoid optional param/*

Reply:
How is to add two classes, `FixtureParamKeyByValue` and `FixtureParamKeyByIndex` that inherit `FixtureParamKey` and represent these two varieties?

Reply:
Two separate classes seems cleaner (I personally prefer the frozen dataclasses to named tuples).

@sadra-barikbin note you can use modern typing annotations now in `main`: `Hashable | None` instead of `Optional[Hashable]`, etc.

Reply:
@nicoddemus , `|` syntax seemingly needs python 3.10 or above. Shall I use it anyway?

Reply:
@RonnyPfannschmidt @nicoddemus ,so shall I add `FixtureArgKey` and two descendent classes `FixtureArgKeyByValue` and `FixtureArgKeyByIndex`?

Reply:
> @nicoddemus , | syntax seemingly needs python 3.10 or above. Shall I use it anyway?

Oh right, we could use `from __future__ import annotations` and use it from Python >=3.7, however this is not without side effects so it is best to just using the legacy syntax, sorry for the noise!

> @RonnyPfannschmidt @nicoddemus ,so shall I add FixtureArgKey and two descendent classes FixtureArgKeyByValue and FixtureArgKeyByIndex?

I think so, or if you prefer and there is not much code to duplicate, you can just have `FixtureArgKeyByValue` and `FixtureArgKeyByIndex` without a base class, using `FixtureArgKey = Union[FixtureArgKeyByValue, FixtureArgKeyByIndex]` for type annotations.

Reply:
:+1: on @nicoddemus's final suggestion - `FixtureArgKey = Union[FixtureArgKeyByValue, FixtureArgKeyByIndex]`